File: ios/GradeThread/Inventory/InventoryStage.swift
import Foundation

/// Stage grouping for the inventory triage tabs — mirrors the web
/// `TO_LIST_STATUSES` / `LISTED_STATUSES` etc. tab structure in
/// `src/pages/flipdesk/listings.tsx`. Each stage maps to a set of
/// canonical `inventory_items.status` values; an item belongs to the
/// stage whose status set contains its current status.
public enum InventoryStage: String, CaseIterable, Identifiable, Hashable {
    case all
    case toList     = "to_list"
    case drafts
    case active
    case sold
    case shipped
    case returned

    public var id: String { rawValue }

    /// Display label shown on the TabView item.
    public var label: String {
        switch self {
        case .all:      return "All"
        case .toList:   return "To list"
        case .drafts:   return "Drafts"
        case .active:   return "Active"
        case .sold:     return "Sold"
        case .shipped:  return "Shipped"
        case .returned: return "Returned"
        }
    }

    /// SF Symbol on the tab item.
    public var systemImage: String {
        switch self {
        case .all:      return "tray.full"
        case .toList:   return "tray.and.arrow.up"
        case .drafts:   return "doc.text"
        case .active:   return "tag"
        case .sold:     return "dollarsign.circle"
        case .shipped:  return "shippingbox"
        case .returned: return "arrow.uturn.backward"
        }
    }

    /// Friendly empty-state message + CTA for when the user has no items
    /// in this stage. Mirrors the web copy.
    public var emptyStateTitle: String {
        switch self {
        case .all:      return "No items yet"
        case .toList:   return "Nothing to list"
        case .drafts:   return "No drafts"
        case .active:   return "No active listings"
        case .sold:     return "No sales yet"
        case .shipped:  return "Nothing shipped yet"
        case .returned: return "No returns"
        }
    }

    public var emptyStateSubtitle: String {
        switch self {
        case .all:      return "Add an item from the + tab to get started."
        case .toList:   return "Catalog photos for an item to move it here."
        case .drafts:   return "Compose a listing draft for any cataloged item."
        case .active:   return "Push a draft to eBay to see it here."
        case .sold:     return "Sold items show here until they ship."
        case .shipped:  return "Items move here once you mark them shipped."
        case .returned: return "Returned sales appear here for reconciliation."
        }
    }

    /// Canonical `inventory_items.status` values that belong to this
    /// stage. Source of truth is the web `listings.tsx` tab predicate.
    public var matchingStatuses: Set<String> {
        switch self {
        case .all:
            return Set(InventoryStage.allKnownStatuses)
        case .toList:
            // Same set as TO_LIST_STATUSES on the web: anything pre-draft
            // that's already past sourcing — items the user could reasonably
            // start listing today.
            // US-3543: `grading` too. An item waiting on its grade is still
            // work to list, and it used to show under All and nowhere else.
            return ["sourced", "acquired", "cataloged", "measured", "photographed", "grading", "graded", "comped"]
        case .drafts:
            return ["drafted"]
        case .active:
            return ["listed"]
        case .sold:
            return ["sold"]
        case .shipped:
            return ["shipped", "completed"]
        case .returned:
            return ["returned"]
        }
    }

    /// US-3543: the order a stage opens in. Sales read newest sale first (the
    /// web's Sold tab default); the work queue reads longest-waiting first, so
    /// the item that has sat the longest is the one on top.
    public var defaultSort: SortOption {
        switch self {
        case .sold, .shipped, .returned: return .recentSale
        case .toList, .drafts:           return .untouchedLongest
        case .all, .active:              return .newest
        }
    }

    /// US-3543: sorts that make sense in this stage. The sale sorts are only
    /// offered where sold items can appear.
    public var sortOptions: [SortOption] {
        switch self {
        case .all, .sold, .shipped, .returned:
            return SortOption.allCases
        case .toList, .drafts, .active:
            return SortOption.allCases.filter { !$0.isSaleSort }
        }
    }

    /// Every status the iOS app knows about. Used by `.all` and for
    /// defensive filtering.
    public static let allKnownStatuses: [String] = [
        "sourced", "acquired", "cataloged", "measured", "photographed",
        "grading", "graded", "comped", "drafted",
        "listed", "sold", "shipped", "completed", "returned",
        "archived", "keeping", "wearing",
    ]

    /// Stages surfaced in the TabView, in display order. Excludes
    /// `archived/keeping/wearing` which aren't worth a top-level tab —
    /// the web hides them too.
    public static let userFacing: [InventoryStage] = [
        .all, .toList, .drafts, .active, .sold, .shipped, .returned,
    ]
}

File: android/app/src/main/java/com/gradethread/app/inventory/InventoryStage.kt
package com.gradethread.app.inventory

import androidx.annotation.StringRes
import com.gradethread.app.R

/**
 * US-1342: the inventory stage tabs (iOS `InventoryStage`).
 *
 * NOTE ON THE ACCEPTANCE CRITERIA: US-1342 describes the tabs as
 * "all/draft/listed/sold/unsold". That does not match the iOS surface this
 * story ports. There are SEVEN stages, "listed" is named `ACTIVE`, and there
 * is no "unsold" stage at all. Parity with iOS is the story's stated purpose,
 * so the iOS set is what's implemented here.
 *
 * US-2976: [wire] is the persisted tab choice and must not change; [label]
 * is a string RESOURCE. They were both plain Strings, which is how the seven
 * tab names along the top of the inventory stayed English in a Spanish app.
 */
enum class InventoryStage(val wire: String, @StringRes val label: Int) {
    ALL("all", R.string.inventory_stage_all),
    TO_LIST("to_list", R.string.inventory_stage_to_list),
    DRAFTS("drafts", R.string.inventory_stage_drafts),
    ACTIVE("active", R.string.inventory_stage_active),
    SOLD("sold", R.string.inventory_stage_sold),
    SHIPPED("shipped", R.string.inventory_stage_shipped),
    RETURNED("returned", R.string.inventory_stage_returned),
    ;

    /** The statuses this stage admits. */
    val matchingStatuses: Set<String>
        get() = when (this) {
            ALL -> allKnownStatuses
            TO_LIST -> setOf(
                "sourced",
                "acquired",
                "cataloged",
                "measured",
                "photographed",
                // US-3543: waiting on a grade is still work to list. It used to
                // show under All and nowhere else.
                "grading",
                "graded",
                "comped",
            )
            DRAFTS -> setOf("drafted")
            ACTIVE -> setOf("listed")
            SOLD -> setOf("sold")
            SHIPPED -> setOf("shipped", "completed")
            RETURNED -> setOf("returned")
        }

    fun matches(status: String): Boolean = status in matchingStatuses

    /**
     * US-3543: the order a stage opens in. Sales read newest sale first (the
     * web's Sold tab default); the work queue reads longest-waiting first.
     */
    val defaultSort: SortOption
        get() = when (this) {
            SOLD, SHIPPED, RETURNED -> SortOption.RECENT_SALE
            TO_LIST, DRAFTS -> SortOption.UNTOUCHED_LONGEST
            ALL, ACTIVE -> SortOption.NEWEST
        }

    /** US-3543: sorts that make sense here; sale sorts only where sales live. */
    val sortOptions: List<SortOption>
        get() = when (this) {
            ALL, SOLD, SHIPPED, RETURNED -> SortOption.entries
            TO_LIST, DRAFTS, ACTIVE -> SortOption.entries.filterNot { it.isSaleSort }
        }

    companion object {
        /**
         * Every status the app knows about — and therefore everything ALL
         * admits, including `archived`/`keeping`/`wearing`, which have no tab
         * of their own.
         */
        val allKnownStatuses: Set<String> = setOf(
            "sourced", "acquired", "cataloged", "measured", "photographed",
            "grading", "graded", "comped", "drafted",
            "listed", "sold", "shipped", "completed", "returned",
            "archived", "keeping", "wearing",
        )

        /** Tab order. */
        val userFacing: List<InventoryStage> = listOf(
            ALL,
            TO_LIST,
            DRAFTS,
            ACTIVE,
            SOLD,
            SHIPPED,
            RETURNED,
        )

        /**
         * Statuses with no tab but All. US-3543 moved `grading` into To list
         * on both platforms; what is left is deliberately off the work tabs.
         */
        val statusesWithoutASpecificTab: Set<String> =
            setOf("archived", "keeping", "wearing")
    }
}
